### ISSUE TITLE
Allow `input` tensor to have arbitrary tailing extra dimensions than `index` tensor in `torch.gather()` / `torch.scatter()`

### ISSUE THREAD

**Issue by XuehaiPan** [pytorch/pytorch#115819]
### 🚀 The feature, motivation and pitch

The [`torch.gather()`](https://pytorch.org/docs/stable/generated/torch.gather.html) function is very useful to select elements in the `input` according to the given `index` tensor.

According to the documentation:

> [`torch.gather(input, dim, index, *, sparse_grad=False, out=None) → Tensor`](https://pytorch.org/docs/stable/tensors.html#torch.Tensor)
>
> Gathers values along an axis specified by dim.
>
> For a 3-D tensor the output is specified by:
>
> ```python
> out[i][j][k] = input[index[i][j][k]][j][k]  # if dim == 0
> out[i][j][k] = input[i][index[i][j][k]][k]  # if dim == 1
> out[i][j][k] = input[i][j][index[i][j][k]]  # if dim == 2
> ```
>
> `input` and `index` must have the same number of dimensions. It is also required that `index.size(d) <= input.size(d)` for all `dimensions d != dim`. `out` will have the same shape as `index`. Note that `input` and `index` do not broadcast against each other.

The `index` tensor needs to have the same dimension as `input`.

------

Now I have a use case that the index has a smaller dimension than the `input` tensor. For example:

```python
input: 3-dim tensor
index: 2-dim tensor

out[i][j][k] = input[index[i][j]][j][k]  # if dim == 0
out[i][j][k] = input[i][index[i][j]][k]  # if dim == 1
```

Currently, I have to work around this by expanding the dimension manually:

```python
import torch

input  # size = (B, L, E)
index  # size = (B, 1)

expanded_index = expanded_index.unsqueeze(dim=-1).expand(-1, -1, input.size(-1))  # size = (B, 1, E)

out = torch.gather(input, dim=1, index=index)  # size = (B, 1, E)
```

It would be nice to automatically broadcast the tailing dimensions from `input` to `index` (the tailing dimension `E` in the above example).

### Alternatives

There is also an alternative to using `bool` masks. However it is not efficient while the tailing dimension is very large.

```python
import torch
import torch.nn.functional as F

input  # size = (B, L, E)
index  # size = (B, 1)

mask = F.one_hot(index, num_classes=input.size(1))  # size = (B, L)
out = input[mask.bool()]  # size = (B, E)
```

**Comment by vadimkantorov** [comment]
Related: 
- https://github.com/pytorch/pytorch/issues/64208

**Comment by the-jb** [comment]
Seems [torch.take_along_dim](https://pytorch.org/docs/stable/generated/torch.take_along_dim.html) is what you are looking for.

```
import torch

input  # size = (B, L, E)
index  # size = (B, 1)

out = torch.take_along_dim(input, dim=1, indices=index.unsqueeze(-1))  # size = (B, 1, E)
```
